lfsapi/client.go

```diff
@@ -0,0 +1,149 @@
+package lfsapi
+
+import (
+	"crypto/tls"
+	"io"
+	"net"
+	"net/http"
+	"net/url"
+	"strings"
+	"time"
+
+	"github.com/git-lfs/git-lfs/errors"
+	"github.com/rubyist/tracerx"
+)
+
+func (c *Client) Do(req *http.Request) (*http.Response, error) {
+	res, err := c.doWithRedirects(c.httpClient(req.Host), req, nil)
+	if err != nil {
+		return res, err
+	}
+
+	return res, c.handleResponse(res)
+}
+
+func (c *Client) doWithRedirects(cli *http.Client, req *http.Request, via []*http.Request) (*http.Response, error) {
+	if seeker, ok := req.Body.(io.Seeker); ok {
+		seeker.Seek(0, io.SeekStart)
+	}
+
+	res, err := cli.Do(req)
+	if err != nil {
+		return res, err
+	}
+
+	if res.StatusCode != 307 {
+		return res, err
+	}
+
+	redirectTo := res.Header.Get("Location")
+	locurl, err := url.Parse(redirectTo)
+	if err == nil && !locurl.IsAbs() {
+		locurl = req.URL.ResolveReference(locurl)
+		redirectTo = locurl.String()
+	}
+
+	via = append(via, req)
+	redirectedReq, err := http.NewRequest(req.Method, redirectTo, nil)
+	if err != nil {
+		return res, errors.Wrapf(err, err.Error())
+	}
+
+	redirectedReq.Body = req.Body
+	redirectedReq.ContentLength = req.ContentLength
+
+	if err = checkRedirect(redirectedReq, via); err != nil {
+		return res, errors.Wrapf(err, err.Error())
```

Comment:
What does `errors.Wrapf(err, err.Error())` do? I think this would just turn an error "foo" into "foo: foo".